src/Optimizer/Transformer/MinifyHtml.php

```diff
@@ -0,0 +1,242 @@
+<?php
+
+namespace AmpProject\Optimizer\Transformer;
+
+use AmpProject\Dom\Document;
+use AmpProject\Dom\Element;
+use AmpProject\Optimizer\Configuration\MinifyHtmlConfiguration;
+use AmpProject\Optimizer\ErrorCollection;
+use AmpProject\Optimizer\Transformer;
+use AmpProject\Optimizer\TransformerConfiguration;
+use DOMComment;
+use DOMNode;
+use DOMText;
+
+/**
+ * Transformer that that minifies HTML.
+ *
+ * @package ampproject/amp-toolbox
+ */
+final class MinifyHtml implements Transformer
+{
+
+    /**
+     * Configuration store to use.
+     *
+     * @var TransformerConfiguration
+     */
+    private $configuration;
+
+    /**
+     * Store nodes for later deletion
+     *
+     * @var array
+     */
+    private $nodesToRemove;
+
+    /**
+     * Instantiate an OptimizeAmpBind object.
+     *
+     * @param TransformerConfiguration $configuration Configuration store to use.
+     */
+    public function __construct(TransformerConfiguration $configuration)
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Apply transformations to the provided DOM document.
+     *
+     * @param Document        $document DOM document to apply the transformations to.
+     * @param ErrorCollection $errors   Collection of errors that are collected during transformation.
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if (! $this->configuration->get(MinifyHtmlConfiguration::MINIFIY)) {
+            return;
+        }
+
+        /*
+         * Store nodes for later deletion to avoid changing the tree structure
+         * while iterating the DOM.
+         */
+        $this->nodesToRemove = [];
+
+        // Recursively walk through all nodes and minify if possible.
+        $this->minifyNode($document, $this->configuration->toArray());
```

Comment:
The nodes to remove can be collected via the return values.
```suggestion
        // Recursively walk through all nodes and minify if possible.
        $nodesToRemove = $this->minifyNode($document, $this->configuration->toArray());
```